awx/main/scheduler/task_manager_models.py

```diff
@@ -64,58 +133,67 @@ def __contains__(self, hostname):
 
 
 class TaskManagerInstanceGroups:
-    """A class representing minimal data the task manager needs to represent an InstanceGroup."""
+    """A class representing minimal data the task manager needs to represent all the InstanceGroups."""
 
-    def __init__(self, instances_by_hostname=None, instance_groups=None, instance_groups_queryset=None):
+    def __init__(self, task_manager_instances=None, instance_groups=None, instance_groups_queryset=None):
         self.instance_groups = dict()
+        self.task_manager_instances = task_manager_instances if task_manager_instances is not None else TaskManagerInstances()
         self.controlplane_ig = None
         self.pk_ig_map = dict()
+        self.control_task_impact = settings.AWX_CONTROL_NODE_TASK_IMPACT
 
         if instance_groups is not None:  # for testing
             self.instance_groups = instance_groups
         else:
             if instance_groups_queryset is None:
-                instance_groups_queryset = InstanceGroup.objects.prefetch_related('instances').only('name', 'instances')
+                instance_groups_queryset = InstanceGroup.objects.prefetch_related('instances').only(
+                    'name', 'instances', 'max_concurrent_jobs', 'max_forks', 'is_container_group'
+                )
             for instance_group in instance_groups_queryset:
                 if instance_group.name == settings.DEFAULT_CONTROL_PLANE_QUEUE_NAME:
                     self.controlplane_ig = instance_group
-                self.instance_groups[instance_group.name] = dict(
-                    instances=[
-                        instances_by_hostname[instance.hostname] for instance in instance_group.instances.all() if instance.hostname in instances_by_hostname
-                    ],
-                )
+                self.instance_groups[instance_group.name] = TaskManagerInstanceGroup(instance_group, self.task_manager_instances)
                 self.pk_ig_map[instance_group.pk] = instance_group
 
+    def __getitem__(self, ig_name):
+        return self.instance_groups.get(ig_name)
+
+    def __contains__(self, ig_name):
+        return ig_name in self.instance_groups
+
     def get_remaining_capacity(self, group_name):
-        instances = self.instance_groups[group_name]['instances']
-        return sum(inst.remaining_capacity for inst in instances)
+        return self.instance_groups[group_name].get_remaining_instance_capacity()
 
     def get_consumed_capacity(self, group_name):
-        instances = self.instance_groups[group_name]['instances']
-        return sum(inst.consumed_capacity for inst in instances)
+        return self.instance_groups[group_name].get_consumed_instance_capacity()
+
+    def consume_capacity(self, task):
+        self.task_manager_instances.consume_capacity(task)
+        if task.instance_group is not None and task.instance_group.is_container_group:
+            self.instance_groups[task.instance_group.name].consume_capacity(task)
```

Comment:
```python
        if task.instance_group_id:
             ig = self.pk_ig_map[task.instance_group_id]
            if ig.is_container_group:
                self.instance_groups[ig.name].consume_capacity(task)
```

for performance, I think this will allow dropping the prefetch of `instance_group`